File: p4/Deepspace/SpaceStation.rb
#encoding:utf-8
require_relative "ShotResult"
require_relative "CardDealer"
require_relative 'SpaceStationToUI'

module Deepspace
	class SpaceStation
		@@not_used = -1
		def initialize(n, supplies)
			@MAXFUEL = 100 
			@SHIELDLOSSPERUNITSHOT = 0.1     
			@name=n
			@nMedals = 0
				
			@hangar = nil
			@pendingDamage = nil
			@shieldBoosters = []  
			@weapons = [] 
			@ammoPower= supplies.ammoPower
			@shieldPower= supplies.shieldPower
			@fuelUnits = supplies.fuelUnits
						
		end

		def self.newCopy(station)
			@MAXFUEL = 100
			@SHIELDLOSSPERUNITSHOT = 0.1
			@name = station.name
			@nMedals = station.nMedals
			@hangar = station.hangar
			@pendingDamage = station.pendingDamage
			@shieldBoosters = station.shieldBoosters
			@weapons = station.weapons
			@ammoPower= station.ammoPower
			@shieldPower= station.shieldPower
			@fuelUnits = station.fuelUnits
		end
		
		def assignFuelValue(f)
			if(f<@MAXFUEL)
				@fuelUnits = f
			else
				@fuelUnits = @MAXFUEL
			end
		end
		
		def cleanPendingDamage()
			if(@pendingDamage.hasNoEffect())
				@pendingDamage = nil
			end
		end
		
		
		attr_reader:ammoPower,:fuelUnits,:hangar,:name,:nMedals,:pendingDamage,:shieldBoosters,:shieldPower,:weapons
		
		def cleanUpMountedItems()
			@weapons.delete_if {|w| w.uses == 0}
			@shieldBoosters.delete_if {|s| s.uses == 0}
		end
		
		def discardHangar()
			@hangar = nil
		end
		
		def discardShieldBooster(i)
			size = @shieldBoosters.count
			if(i >= 0 && i < size)
				s = @shieldBoosters.delete_at(i)
				if(@pendingDamage != nil)
					discardShield(s)
					cleanPendingDamage()
				end
			end
			
		end
		
		def discardShieldBoosterInHangar(i)
			if(@hangar != nil)
				@hangar.removeShieldBooster(i)
			end
		end
		
		def discardWeapon(i)
			size = @weapons.size()
			if(i >= 0 && i < size)
				w = @weapons.delete_at(i)
				if(@pendingDamage != nil)
					discardWeapon(w)
					cleanPendingDamage()
				end
			end
		end
		
		def discardWeaponInHangar(i)
			if(@hangar != nil)
				@hangar.removeWeapon(i)
			end
		end
		
		def fire()
			factor=1
        
            for w in @weapons do
                factor*=w.useIt()
            end
            
            return @ammoPower*factor
		end
		
		def speed()
			return @fuelUnits/@MAXFUEL
		end
		
		def getUIversion()
			SpaceStationToUI.new(self)
		end
		
		def mountShieldBooster(i)
			if (@hangar != nil && i!=@@not_used)
                aux = @hangar.removeShieldBooster(i)
                if (aux != nil)
                    @shieldBoosters.push(aux)
                end 
            end 
		end
		
		def mountWeapon(i)
			if (@hangar != nil && i!=@@not_used)
                aux = @hangar.removeWeapon(i)
                if (aux != nil)
                    @weapons.push(aux)
                end 
            end 
		end
		
		def move()
			if (@fuelUnits -  speed() >= 0)
                @fuelUnits -=  speed()
            else 
                @fuelUnits = 0
            end
		end
		
		def protection()
			factor=1
        
            for s in @shieldBoosters do
                factor*=s.useIt()
            end
            
            return @shieldPower*factor
		end
		
		def receiveHangar(h)
			if(@hangar == nil)
				@hangar = h
			end
		end
		
		def receiveShieldBooster(s)
			if(@hangar == nil)
				return false
			else
				return @hangar.addShieldBooster(s)
			end
		end
		
		def receiveShot(shot)
			myProtection= protection()
            if(myProtection>=shot)
                remaining=@SHIELDLOSSPERUNITSHOT*shot
                if (@shieldPower-remaining>=0)
                    @shieldPower = @shieldPower-remaining
                else 
                    @shieldPower = 0
                end                
                result= ShotResult::RESIST
            else
                result= ShotResult::DONOTRESIST
            end
            return result
		end
		
		def receiveSupplies(s)
			@fuelUnits += s.fuelUnits
			@ammoPower += s.ammoPower
			@shieldPower += s.shieldPower
		end
		
		def receiveWeapon(w)
			if(@hangar == nil)
				return false
			else
				return @hangar.addWeapon(w)
			end
		end
		
		def setLoot(loot)
			dealer = CardDealer.instance()
            h = loot.nHangars()
            
            if (h>0)
                hangar=dealer.nextHangar()
                receiveHangar(hangar)
            end
            
            elements= loot.nSupplies()

            elements.times do 
                sup= dealer.nextSuppliesPackage()
                receiveSupplies(sup)
            end
            
            elements= loot.nWeapons()
            
            elements.times do 
                weap=dealer.nextWeapon()
                receiveWeapon(weap)
            end
            
            elements= loot.nShields()
            
            elements.times do 
                sh= dealer.nextShieldBooster()
                receiveShieldBooster(sh)
            end
            
            medals=loot.nMedals()
            @nMedals+=medals
		end
		
		def setPendingDamage(d)
			@pendingDamage = d.adjust(@weapons,@shieldBoosters)
		end
		
		def validState()
			return (@pendingDamage == nil || @pendingDamage.hasNoEffect())
		end
		
		def to_s()
            getUIversion().to_s
        end 
	
	
	end
end			

File: p5/deepspace/src/deepspace/SpaceStation.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package deepspace;
import java.util.ArrayList;

/**
 *
 * @author cristobalmer
 */
class SpaceStation implements SpaceFighter{
    private static final float MAXFUEL = 100f;
    private static final float SHIELDLOSSPERUNITSHOT = 0.1f;
    
    private float ammoPower;
    private float fuelUnits;
    private String name;
    private int nMedals;
    private float shieldPower;
    
    private ArrayList<Weapon> weapons;
    private ArrayList<ShieldBooster> shieldBoosters;
    private Hangar hangar;
    private Damage pendingDamage;
    
    private void assignFuelValue(float f){
        if (f<MAXFUEL){
            fuelUnits = f;
        } else fuelUnits = MAXFUEL;
    }
    
    private void cleanPendingDamage(){
        if (pendingDamage.hasNoEffect()){
            pendingDamage = null;
        }
    }
    
    SpaceStation(String n,SuppliesPackage supplies){
        nMedals = 0;
        weapons = new ArrayList<>();
        shieldBoosters = new ArrayList<>();
        name = n;
        receiveSupplies(supplies);
        hangar = null;
        pendingDamage = null;
        
    }
    
    SpaceStation(SpaceStation station){
        ammoPower = station.ammoPower;
        fuelUnits = station.fuelUnits;
        name = station.name;
        nMedals = station.nMedals;
        shieldPower = station.shieldPower;

        if ( station.pendingDamage != null )
            pendingDamage = station.pendingDamage.copy();
        if ( station.weapons != null )
            weapons = new ArrayList<>(station.weapons);
        if ( station.shieldBoosters != null )
            shieldBoosters = new ArrayList<>(station.shieldBoosters);
        if ( station.hangar != null )
            hangar = new Hangar(station.hangar);  
    }
    
    
    public void cleanUpMountedItems(){
       weapons.removeIf(w->(w.getUses() == 0));
       shieldBoosters.removeIf(s->(s.getUses() == 0));
    }
    
    public void discardHangar(){
        hangar = null;
    }
    
    public void discardShieldBooster(int i){
        int size = shieldBoosters.size();
        if(i>=0 && i<size){
            ShieldBooster s = shieldBoosters.remove(i);
            if(pendingDamage != null){
                pendingDamage.discardShieldBooster();
                cleanPendingDamage();
            }        
        }
    }
    
    public void discardShieldBoosterInHangar(int i){
        if (hangar != null){
            hangar.removeShieldBooster(i);
        }
    }
    
    public void discardWeapon(int i){
        int size = weapons.size();
        if(i>=0 && i<size){
            Weapon w =weapons.remove(i);
            if(pendingDamage != null){
                pendingDamage.discardWeapon(w);
                cleanPendingDamage();
            }        
        }
    }
    
    public void discardWeaponInHangar(int i){
        if (hangar != null){
            hangar.removeWeapon(i);
        }
    }
    
    @Override
    public float fire(){
        int size = weapons.size();
        float factor = 1;
        Weapon w;
        for (int i =0;i<size;i++){
            w = weapons.get(i);
            factor *= w.useIt();
        }
        return ammoPower*factor;
    }
    
    public float getAmmoPower(){
        return ammoPower;
    }
    
    public float getFuelUnits(){
        return fuelUnits;
    }
    
    public Hangar getHangar(){
        return hangar;
    }
    
    public String getName(){
        return name;
    }
    
    public int getNMedals(){
        return nMedals;
    }
    
    public Damage getPendingDamage(){
        return pendingDamage;
    }
    
    public ArrayList<ShieldBooster> getShieldBoosters(){
        return shieldBoosters;
    }
    
    public float getShieldPower(){
        return shieldPower;
    }
    
    public float getSpeed(){
        return fuelUnits/MAXFUEL;
    }
    
    public SpaceStationToUI getUIversion(){
        return new SpaceStationToUI(this);
    }
    
    public ArrayList<Weapon> getWeapons(){
        return weapons;
    }
    
    public void mountShieldBooster(int i){
        if (hangar != null && i!=-1){
            ShieldBooster aux = hangar.removeShieldBooster(i);
            if (aux != null){
                shieldBoosters.add(aux);
            }
        }
    }
    
    public void mountWeapon(int i){
        if (hangar != null && i!=-1){
            Weapon aux = hangar.removeWeapon(i);
            if (aux != null){
                weapons.add(aux);
            }
        }
    }
    
    public void move(){
        if (fuelUnits - fuelUnits*getSpeed()>=0){
            fuelUnits -= fuelUnits*getSpeed();
        }
    }
    
    @Override
    public float protection(){
        int size = shieldBoosters.size();
        float factor = 1;
        ShieldBooster s;
        for (int i =0;i<size;i++){
            s = shieldBoosters.get(i);
            factor *= s.useIt();
        }
        return shieldPower*factor;
    }
            
    public void receiveHangar(Hangar h){
        if (hangar == null){
            hangar = h;
        }
    }
    
    public boolean receiveShieldBooster(ShieldBooster s){
        boolean out = false;
        if (hangar!=null){
            out = hangar.addShieldBooster(s);
        }
         return out;        
    }
    
    @Override
    public ShotResult receiveShot(float shot){
        ShotResult out;
        float myProtection = protection();
        if(myProtection>=shot){
            shieldPower-=SHIELDLOSSPERUNITSHOT *shot;
            if (shieldPower < 0.0f){
                shieldPower = 0.0f;
            }
            out = ShotResult.RESIST;
        }else{
            shieldPower=0.0f;
            out = ShotResult.DONOTRESIST;
        }
        return out;
    }
    
    
    public void receiveSupplies(SuppliesPackage s){
        ammoPower += s.getAmmoPower();
        assignFuelValue(fuelUnits+s.getFuelUnits());
        shieldPower += s.getShieldPower();
    }
    
    public boolean receiveWeapon(Weapon w){
        boolean out = false;
        if (hangar!=null){
            out = hangar.addWeapon(w);
        } 
         return out;
    }
    
    public Transformation setLoot(Loot loot){
        CardDealer dealer = CardDealer.getInstance();
        int h = loot.getNHangars();
        if(h>0){
            Hangar hangar = dealer.nextHangar();
            receiveHangar(hangar);
        }
        
        int elements = loot.getNSupplies();
        for(int i =0;i<elements;i++){
            SuppliesPackage sup = dealer.nextSuppliesPackage();
            receiveSupplies(sup);
        }
        
        elements = loot.getNWeapons();
         for(int i =0;i<elements;i++){
            Weapon weap = dealer.nextWeapon();
            receiveWeapon(weap);
        }
         
         elements = loot.getNShields();
         for(int i =0;i<elements;i++){
            ShieldBooster sh =dealer.nextShieldBooster();
            receiveShieldBooster(sh);
        }
        
         int medals = loot.getNMedals();
         nMedals+=medals;
         
         if(loot.getEfficient()){
            return Transformation.GETEFFICIENT;
        }else if(loot.spaceCity()){
            return Transformation.SPACECITY;
        }else{
            return Transformation.NOTRANSFORM;
        }
    }
    
    public void setPendingDamage(Damage d){
        pendingDamage = d.adjust(weapons, shieldBoosters);
    }
    
    public boolean validState(){
        return (pendingDamage == null || pendingDamage.hasNoEffect());
    }
    
    
}